core/registration.c

```diff
@@ -191,30 +199,102 @@ static void prv_handleRegistrationReply(lwm2m_transaction_t * transacP,
+
+    if (contextP != NULL) targetP = utils_findServer(contextP, transacP->peerH);
 
-    if (targetP->status == STATE_REG_PENDING)
+    if (targetP != NULL && targetP->status == STATE_REG_PENDING)
     {
         time_t tv_sec = lwm2m_gettime();
         if (tv_sec >= 0)
         {
             targetP->registration = tv_sec;
         }
-        if (packet != NULL && packet->code == COAP_201_CREATED)
+        /* handle reply for block transfer of registration request */
+        if (packet != NULL && IS_OPTION(packet, COAP_OPTION_BLOCK1))
         {
-            targetP->status = STATE_REGISTERED;
-            if (NULL != targetP->location)
+            /* we should send next blocks only if we get success */
+            if ((packet->code & 0xE0) == 2 << 5)
             {
-                lwm2m_free(targetP->location);
+                uint32_t block1_num;
+                uint8_t  block1_more;
+                uint16_t block1_size;
+                uint32_t block1_offset;
+                coap_get_header_block1(packet, &block1_num, &block1_more, &block1_size, &block1_offset);
+                targetP->rem_payload_length -= block1_size;
+                if (block1_more == 0) targetP->final_block_response = true;
+
+                if (packet != NULL && (packet->code == COAP_201_CREATED || packet->code == COAP_231_CONTINUE))
+                {
+                    if (packet->code == COAP_201_CREATED)
+                    {
+                        if (NULL != targetP->location)
+                        {
+                            lwm2m_free(targetP->location);
+                        }
+                        targetP->location = coap_get_multi_option_as_string(packet->location_path);
+                    }
+                    /* All blocks have been sent */
+                    if (targetP->rem_payload_length <= 0)
+                    {
+                        /* registration is successful only if we had got a final response in
+                         * any of the block transfer requests */
+                        if (targetP->final_block_response)
+                        {
+                            targetP->status = STATE_REGISTERED;
+                            LOG("Registration successful");
+                        }
+                        else
+                        {
+                            targetP->status = STATE_REG_FAILED;
+                            LOG("Registration failed");
+                        }
+                        targetP->block_no = 0;
+                        targetP->block_size = 0;
+                        targetP->rem_payload_length = 0;
+                        targetP->final_block_response = false;
+                    }
+                }
+                if (targetP->rem_payload_length > 0)
+                {
+                    if (block1_size <= targetP->block_size)
+                    {
+                        targetP->block_no += targetP->block_size/block1_size;
```

Comment:
Missing targetP->block_no update when block1_size > targetP->block_size. Should just go to the next block number.